Can't open local folder on Android

**Describe the bug**
When I grant storage permissions and try to add an empty folder, nothing happens.

**To Reproduce**
Steps to reproduce the behavior:
1. Install latest nightly version of Logseq on Android
2. Click on 'Grant permission' and add an empty folder as a new graph

**Expected behavior**
Adding an empty folder generates a new graph.

**Smartphone:**
 - Device: Realme 5i
 - OS: Android 10
 - Browser: Chrome
 - Version: Android Nightly Release 20220106

The same issue on my tablet

Mi Pad 4 Plus
Android 8.1.0
MIUI by xiaomi.eu 10.3

Vault path: `/storage/emulated/0/Documents/Logseq/logseq-vault/`

Stacktrace:

```
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
 at com.getcapacitor.Bridge.lambda$callPluginMethod$0$Bridge(Bridge.java:601)
 at com.getcapacitor.-$$Lambda$Bridge$25SFHybyAQk7zS27hTVXh2p8tmw.run(Unknown Source:8)
 at android.os.Handler.handleCallback(Unknown Source:2)
 at android.os.Handler.dispatchMessage(Unknown Source:4)
 at android.os.Looper.loop(Unknown Source:152)
 at android.os.HandlerThread.run(Unknown Source:28)
Caused by: java.lang.reflect.InvocationTargetException
 at java.lang.reflect.Method.invoke(Native Method)
 at com.getcapacitor.PluginHandle.invoke(PluginHandle.java:121)
 at com.getcapacitor.Bridge.lambda$callPluginMethod$0$Bridge(Bridge.java:592)
 ... 5 more
Caused by: java.lang.NullPointerException: uriString
 at android.net.Uri$StringUri.<init>(Unknown Source:22)
 at android.net.Uri$StringUri.<init>(Unknown Source:0)
 at android.net.Uri.parse(Unknown Source:3)
 at com.capacitorjs.plugins.filesystem.Filesystem.getFileObject(Filesystem.java:202)
 at com.capacitorjs.plugins.filesystem.Filesystem.readdir(Filesystem.java:77)
 at com.capacitorjs.plugins.filesystem.FilesystemPlugin.readdir(FilesystemPlugin.java:270)
 ... 8 more
```

I have now experienced this issue on multiple phones, on Android 10, 11, and 12, across multiple versions of Logseq from 0.6.2 up to the latest nightly. Am currently experiencing it on a Google Pixel 5, where adding a graph worked initially but mysteriously stopped working after a reinstall.

I made an attempt at fishing out the error message from the crash through logcat:

```
FATAL EXCEPTION: CapacitorPlugins
Process: com.logseq.app, PID: 7908
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at com.getcapacitor.Bridge.lambda$callPluginMethod$0$Bridge(Bridge.java:601)
	at com.getcapacitor.-$$Lambda$Bridge$25SFHybyAQk7zS27hTVXh2p8tmw.run(Unknown Source:8)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.os.HandlerThread.run(HandlerThread.java:67)
Caused by: java.lang.reflect.InvocationTargetException
	at java.lang.reflect.Method.invoke(Native Method)
	at com.getcapacitor.PluginHandle.invoke(PluginHandle.java:121)
	at com.getcapacitor.Bridge.lambda$callPluginMethod$0$Bridge(Bridge.java:592)
	... 6 more
Caused by: android.content.ActivityNotFoundException: No Activity found to handle Intent { act=android.intent.action.OPEN_DOCUMENT_TREE cat=[android.intent.category.DEFAULT] }
	at android.app.Instrumentation.checkStartActivityResult(Instrumentation.java:2117)
	at android.app.Instrumentation.execStartActivity(Instrumentation.java:1762)
	at android.app.Activity.startActivityForResult(Activity.java:5407)
	at androidx.activity.ComponentActivity.startActivityForResult(ComponentActivity.java:584)
	at androidx.core.app.ActivityCompat.startActivityForResult(ActivityCompat.java:234)
	at androidx.activity.ComponentActivity$2.onLaunch(ComponentActivity.java:218)
	at androidx.activity.result.ActivityResultRegistry$2.launch(ActivityResultRegistry.java:162)
	at androidx.activity.result.ActivityResultLauncher.launch(ActivityResultLauncher.java:47)
	at com.getcapacitor.Plugin.startActivityForResult(Plugin.java:181)
	at com.logseq.app.FolderPicker.pickFolder(FolderPicker.java:40)
	... 9 more

> I have now experienced this issue on multiple phones, on Android 10, 11, and 12, across multiple versions of Logseq from 0.6.2 up to the latest nightly. Am currently experiencing it on a Google Pixel 5, where adding a graph worked initially but mysteriously stopped working after a reinstall. It takes place with empty graphs, and with large graphs, and after removing special characters from file names. It also takes place with graphs placed in the Android home folder and in subdirectories.
> 
> I made an attempt at fishing out the error message from the crash through logcat. It appears to be the same as the one shared by @MrModest
> 
> ```
> FATAL EXCEPTION: CapacitorPlugins
> Process: com.logseq.app, PID: 7908
> java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
> 	at com.getcapacitor.Bridge.lambda$callPluginMethod$0$Bridge(Bridge.java:601)
> 	at com.getcapacitor.-$$Lambda$Bridge$25SFHybyAQk7zS27hTVXh2p8tmw.run(Unknown Source:8)
> 	at android.os.Handler.handleCallback(Handler.java:938)
> 	at android.os.Handler.dispatchMessage(Handler.java:99)
> 	at android.os.Looper.loopOnce(Looper.java:201)
> 	at android.os.Looper.loop(Looper.java:288)
> 	at android.os.HandlerThread.run(HandlerThread.java:67)
> Caused by: java.lang.reflect.InvocationTargetException
> 	at java.lang.reflect.Method.invoke(Native Method)
> 	at com.getcapacitor.PluginHandle.invoke(PluginHandle.java:121)
> 	at com.getcapacitor.Bridge.lambda$callPluginMethod$0$Bridge(Bridge.java:592)
> 	... 6 more
> Caused by: android.content.ActivityNotFoundException: No Activity found to handle Intent { act=android.intent.action.OPEN_DOCUMENT_TREE cat=[android.intent.category.DEFAULT] }
> 	at android.app.Instrumentation.checkStartActivityResult(Instrumentation.java:2117)
> 	at android.app.Instrumentation.execStartActivity(Instrumentation.java:1762)
> 	at android.app.Activity.startActivityForResult(Activity.java:5407)
> 	at androidx.activity.ComponentActivity.startActivityForResult(ComponentActivity.java:584)
> 	at androidx.core.app.ActivityCompat.startActivityForResult(ActivityCompat.java:234)
> 	at androidx.activity.ComponentActivity$2.onLaunch(ComponentActivity.java:218)
> 	at androidx.activity.result.ActivityResultRegistry$2.launch(ActivityResultRegistry.java:162)
> 	at androidx.activity.result.ActivityResultLauncher.launch(ActivityResultLauncher.java:47)
> 	at com.getcapacitor.Plugin.startActivityForResult(Plugin.java:181)
> 	at com.logseq.app.FolderPicker.pickFolder(FolderPicker.java:40)
> 	... 9 more
> ```

Update: I figured out what was causing. On both my phones, I'd uninstalled the stock file manager app and replaced it with a third party file manager. The intent probably only plugins into the stock file manager, and when it isn't present, gives an error. It's possible that the reason for this crash on other phones is that the manufacturer OS doesn't ship the AOSP file manager/file picker that is required by this intent.

Interesting. When I tried to choose folder like `/storage/emulated/0/Logseq` (or `Download` folder, I got a new stacktrace:

```
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at com.getcapacitor.Bridge.lambda$callPluginMethod$0$Bridge(Bridge.java:601)
	at com.getcapacitor.-$$Lambda$Bridge$25SFHybyAQk7zS27hTVXh2p8tmw.run(Unknown Source:8)
	at android.os.Handler.handleCallback(Unknown Source:2)
	at android.os.Handler.dispatchMessage(Unknown Source:4)
	at android.os.Looper.loop(Unknown Source:152)
	at android.os.HandlerThread.run(Unknown Source:28)
Caused by: java.lang.reflect.InvocationTargetException
	at java.lang.reflect.Method.invoke(Native Method)
	at com.getcapacitor.PluginHandle.invoke(PluginHandle.java:121)
	at com.getcapacitor.Bridge.lambda$callPluginMethod$0$Bridge(Bridge.java:592)
	... 5 more
Caused by: java.lang.NoSuchMethodError: No direct method <init>(Ljava/io/File;I)V in class Landroid/os/FileObserver; or its super classes (declaration of 'android.os.FileObserver' appears in /system/framework/framework.jar)
	at com.logseq.app.FsWatcher$SingleFileObserver.<init>(FsWatcher.java:203)
	at com.logseq.app.FsWatcher.watch(FsWatcher.java:60)
	... 8 more
```

But for `Document` folder error still the previous one

Looks like a bug in capacitors file picker API wrapper, but I didn't find any related report in their GitHub repository... 🤔 

```
Caused by: java.lang.NullPointerException: uriString
        at android.net.Uri$StringUri.<init>(Uri.java:507)
        at android.net.Uri$StringUri.<init>(Uri.java:497)
        at android.net.Uri.parse(Uri.java:469)
        at com.capacitorjs.plugins.filesystem.Filesystem.getFileObject(Filesystem.java:202)
        at com.capacitorjs.plugins.filesystem.Filesystem.readdir(Filesystem.java:77)
        at com.capacitorjs.plugins.filesystem.FilesystemPlugin.readdir(FilesystemPlugin.java:270)
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.getcapacitor.PluginHandle.invoke(PluginHandle.java:121) 
        at com.getcapacitor.Bridge.lambda$callPluginMethod$0$Bridge(Bridge.java:592) 
        at com.getcapacitor.-$$Lambda$Bridge$25SFHybyAQk7zS27hTVXh2p8tmw.run(Unknown Source:8) 
        at android.os.Handler.handleCallback(Handler.java:938) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loopOnce(Looper.java:201) 
        at android.os.Looper.loop(Looper.java:288) 
        at android.os.HandlerThread.run(HandlerThread.java:67) 
```

Crash log for P30 HarmonyOS

```
08-09 20:52:48.655 W/System.err( 8121): java.lang.reflect.InvocationTargetException
08-09 20:52:48.655 W/System.err( 8121): 	at java.lang.reflect.Method.invoke(Native Method)
08-09 20:52:48.655 W/System.err( 8121): 	at com.getcapacitor.Plugin.triggerActivityCallback(Plugin.java:154)
08-09 20:52:48.656 W/System.err( 8121): 	at com.getcapacitor.Plugin.lambda$initializeActivityLaunchers$0$Plugin(Plugin.java:116)
08-09 20:52:48.656 W/System.err( 8121): 	at com.getcapacitor.-$$Lambda$Plugin$9tit0PZrRMA5UA4Z_xqI4yDk4JY.onActivityResult(Unknown Source:6)
08-09 20:52:48.656 W/System.err( 8121): 	at androidx.activity.result.ActivityResultRegistry$1.onStateChanged(ActivityResultRegistry.java:145)
08-09 20:52:48.656 E/AwareLog( 1570): AwareMem_MemoryUtils: trim Memory illegal arg exception, proc:8457 userId:-2 level:40 fromAware:true
08-09 20:52:48.656 W/System.err( 8121): 	at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.java:354)
08-09 20:52:48.656 W/System.err( 8121): 	at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.java:265)
08-09 20:52:48.656 W/System.err( 8121): 	at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.java:307)
08-09 20:52:48.656 W/System.err( 8121): 	at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.java:148)
08-09 20:52:48.656 W/System.err( 8121): 	at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.java:134)
08-09 20:52:48.656 W/System.err( 8121): 	at androidx.lifecycle.ReportFragment.dispatch(ReportFragment.java:68)
08-09 20:52:48.656 I/HiSuggestion_13.0.1.302_IntelligentProvider(29592): Provider.call: getPrefBoolean
08-09 20:52:48.656 W/System.err( 8121): 	at androidx.lifecycle.ReportFragment$LifecycleCallbacks.onActivityPostStarted(ReportFragment.java:187)
08-09 20:52:48.656 W/System.err( 8121): 	at android.app.Activity.dispatchActivityPostStarted(Activity.java:1307)
08-09 20:52:48.656 W/System.err( 8121): 	at android.app.Activity.performStart(Activity.java:8277)
08-09 20:52:48.656 W/System.err( 8121): 	at android.app.ActivityThread.handleStartActivity(ActivityThread.java:4084)
08-09 20:52:48.656 W/System.err( 8121): 	at android.app.servertransaction.TransactionExecutor.performLifecycleSequence(TransactionExecutor.java:235)
08-09 20:52:48.656 W/System.err( 8121): 	at android.app.servertransaction.TransactionExecutor.cycleToPath(TransactionExecutor.java:215)
08-09 20:52:48.657 W/System.err( 8121): 	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:187)
08-09 20:52:48.657 W/System.err( 8121): 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:105)
08-09 20:52:48.657 W/System.err( 8121): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2613)
08-09 20:52:48.657 W/System.err( 8121): 	at android.os.Handler.dispatchMessage(Handler.java:110)
08-09 20:52:48.657 W/System.err( 8121): 	at android.os.Looper.loop(Looper.java:219)
08-09 20:52:48.657 W/System.err( 8121): 	at android.app.ActivityThread.main(ActivityThread.java:8668)
08-09 20:52:48.657 W/System.err( 8121): 	at java.lang.reflect.Method.invoke(Native Method)
08-09 20:52:48.657 W/System.err( 8121): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513)
08-09 20:52:48.657 W/System.err( 8121): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1109)
08-09 20:52:48.657 W/System.err( 8121): Caused by: java.lang.NumberFormatException: For input string: "raw:/storage/emulated/0/Download/asdfasf"
08-09 20:52:48.657 W/System.err( 8121): 	at java.lang.Long.parseLong(Long.java:594)
08-09 20:52:48.657 W/System.err( 8121): 	at java.lang.Long.valueOf(Long.java:808)
08-09 20:52:48.657 W/System.err( 8121): 	at com.logseq.app.FileUtil.getPath(FileUtil.java:39)
08-09 20:52:48.657 W/System.err( 8121): 	at com.logseq.app.FolderPicker.folderPickerResult(FolderPicker.java:59)
08-09 20:52:48.657 W/System.err( 8121): 	... 26 more
```

Device: Xiaomi Mi Pad 4 Plus
Logseq path: `/storage/emulated/0/Documents/Logseq/logseq-vault/` (Internal storage)

> https://nightly.link/logseq/logseq/actions/runs/2829456996

```
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at com.getcapacitor.Bridge.lambda$callPluginMethod$0$Bridge(Bridge.java:601)
	at com.getcapacitor.-$$Lambda$Bridge$25SFHybyAQk7zS27hTVXh2p8tmw.run(Unknown Source:8)
	at android.os.Handler.handleCallback(Unknown Source:2)
	at android.os.Handler.dispatchMessage(Unknown Source:4)
	at android.os.Looper.loop(Unknown Source:152)
	at android.os.HandlerThread.run(Unknown Source:28)
Caused by: java.lang.reflect.InvocationTargetException
	at java.lang.reflect.Method.invoke(Native Method)
	at com.getcapacitor.PluginHandle.invoke(PluginHandle.java:121)
	at com.getcapacitor.Bridge.lambda$callPluginMethod$0$Bridge(Bridge.java:592)
	... 5 more
Caused by: java.lang.NullPointerException: uriString
	at android.net.Uri$StringUri.<init>(Unknown Source:22)
	at android.net.Uri$StringUri.<init>(Unknown Source:0)
	at android.net.Uri.parse(Unknown Source:3)
	at com.capacitorjs.plugins.filesystem.Filesystem.getFileObject(Filesystem.java:202)
	at com.capacitorjs.plugins.filesystem.Filesystem.readdir(Filesystem.java:77)
	at com.capacitorjs.plugins.filesystem.FilesystemPlugin.readdir(FilesystemPlugin.java:270)
	... 8 more
```

> https://nightly.link/logseq/logseq/actions/runs/2829884048

Error popup:
![image](https://user-images.githubusercontent.com/13031058/184010371-303916f1-0df0-4aa1-9369-5df33e14daba.png)

and stacktrace then:

```
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at com.getcapacitor.Bridge.lambda$callPluginMethod$0$Bridge(Bridge.java:601)
	at com.getcapacitor.-$$Lambda$Bridge$25SFHybyAQk7zS27hTVXh2p8tmw.run(Unknown Source:8)
	at android.os.Handler.handleCallback(Unknown Source:2)
	at android.os.Handler.dispatchMessage(Unknown Source:4)
	at android.os.Looper.loop(Unknown Source:152)
	at android.os.HandlerThread.run(Unknown Source:28)
Caused by: java.lang.reflect.InvocationTargetException
	at java.lang.reflect.Method.invoke(Native Method)
	at com.getcapacitor.PluginHandle.invoke(PluginHandle.java:121)
	at com.getcapacitor.Bridge.lambda$callPluginMethod$0$Bridge(Bridge.java:592)
	... 5 more
Caused by: java.lang.NullPointerException: uriString
	at android.net.Uri$StringUri.<init>(Unknown Source:22)
	at android.net.Uri$StringUri.<init>(Unknown Source:0)
	at android.net.Uri.parse(Unknown Source:3)
	at com.capacitorjs.plugins.filesystem.Filesystem.getFileObject(Filesystem.java:202)
	at com.capacitorjs.plugins.filesystem.Filesystem.readdir(Filesystem.java:77)
	at com.capacitorjs.plugins.filesystem.FilesystemPlugin.readdir(FilesystemPlugin.java:270)
	... 8 more
```

--- a/android/app/src/main/java/com/logseq/app/FileUtil.java
+++ b/android/app/src/main/java/com/logseq/app/FileUtil.java
@@ -14,8 +14,10 @@ import android.util.Log;
 
 import java.io.File;
 
+// The following refs
 // https://stackoverflow.com/questions/29713587/how-to-get-the-real-path-with-action-open-document-tree-intent
 // https://gist.github.com/asifmujteba/d89ba9074bc941de1eaa#file-asfurihelper
+// with bug fixes and patches.
 public class FileUtil {
     @TargetApi(Build.VERSION_CODES.KITKAT)
     public static String getPath(final Context context, final Uri uri) {
@@ -29,15 +31,18 @@ public class FileUtil {
                 final String docId = DocumentsContract.getDocumentId(uri);
                 final String[] split = docId.split(":");
                 final String type = split[0];
-
-                if ("primary".equalsIgnoreCase(type)) {
-                    return Environment.getExternalStorageDirectory() + "/" + split[1];
-                }
-
                 // NOTE: It's not a good idea to use storage root as Graph root.
                 String remain = "";
                 if (split.length == 2) {
                     remain = split[1];
+                }
+
+                if ("primary".equalsIgnoreCase(type)) {
+                    return Environment.getExternalStorageDirectory() + "/" + remain;
+                } else if ("home".equalsIgnoreCase(type)) {
+                    return Environment.getExternalStorageDirectory() + "/Documents/" + remain;
+                } else if ("downloads".equalsIgnoreCase(type)) {
+                    return Environment.getExternalStorageDirectory() + "/Download/" + remain; // No 's' here
                 }
 
                 File dir = null;

